qiskit/pulse/transforms/dag.py

```diff
@@ -0,0 +1,99 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""A collection of functions to convert ScheduleBlock to DAG representation."""
+
+import retworkx as rx
+
+from qiskit.pulse.schedule import ScheduleBlock
+
+
+def block_to_dag(block: ScheduleBlock) -> rx.PyDAG:
+    """Convert schedule block instruction into DAG.
+
+    ``ScheduleBlock`` can be represented as a DAG as needed.
+    For example, equality of two programs are efficiently checked on DAG representation.
+
+    .. code-block:: python
+
+        with pulse.build() as sched1:
+            with pulse.align_left():
+                pulse.play(my_gaussian0, pulse.DriveChannel(0))
+                pulse.shift_phase(1.57, pulse.DriveChannel(2))
+                pulse.play(my_gaussian1, pulse.DriveChannel(1))
+
+        with pulse.build() as sched2:
+            with pulse.align_left():
+                pulse.shift_phase(1.57, pulse.DriveChannel(2))
+                pulse.play(my_gaussian1, pulse.DriveChannel(1))
+                pulse.play(my_gaussian0, pulse.DriveChannel(0))
+
+    Here the ``sched1 `` and ``sched2`` are different implementations of the same program,
+    but it is difficult to confirm on the list representation.
+
+    Another example is instruction optimization.
+
+    .. code-block:: python
+
+        with pulse.build() as sched:
+            with pulse.align_left():
+                pulse.shift_phase(1.57, pulse.DriveChannel(1))
+                pulse.play(my_gaussian0, pulse.DriveChannel(0))
+                pulse.shift_phase(-1.57, pulse.DriveChannel(1))
+
+    In above program two ``shift_phase`` instructions can be cancelled out because
+    they are consecutive on the same drive channel.
+    This can be easily found on the DAG representation.
+
+    Args:
+        block: A schedule block to be converted.
+
+    Returns:
+        Instructions in DAG representation.
+    """
+    if block.alignment_context.is_sequential:
+        return _sequential_allocation(block)
+    else:
+        return _parallel_allocation(block)
+
+
+def _sequential_allocation(block: ScheduleBlock) -> rx.PyDAG:
+    """A helper function to create a DAG of a sequential alignment context."""
+    dag_instructions = rx.PyDAG()
+
+    prev_node = None
+    edges = []
+    for inst in block.instructions:
```

Comment:
Yes, it can contain another block with different policy, for example

```python
with build() as block:
    with align_left():
        with align_sequential():
            play(my_pulse, d0)        
```

Reply:
Where is the recursive call? My eyes aren't catching it...

Reply:
More explicitly,
```python
block1 = ScheduleBlock(alignment_context=AlignLeft())
block2 = ScheduleBlock(alignment_context=AlignSequential())
block2.append(Play(my_pulse, d0))

block1.append(block2)
```
here the `block2` is nested to `block1`.

Reply:
These alignments are processed in `qiskit.pulse.transforms.canonicalization.block_to_schedule`